File: bindings/swift/Tests/Itb3Tests/RuntimeTests.swift
/*
 * Runtime diagnostics surface: GOMAXPROCS query / set / restore, the
 * heap-profile writer, the pool-counter snapshot and its slot layout,
 * the hash-registry enumeration, and the diagnostic's survival of a
 * text wider than the relay layer's inline buffer.
 */

import Foundation
import Glibc
import XCTest
@testable import Itb3

final class RuntimeTests: XCTestCase {
    /// Zero queries, a positive value sets and returns the previous
    /// one; the query never changes the setting.
    func testGOMAXPROCS() {
        let orig = ItbRuntime.setGOMAXPROCS(0)
        XCTAssertGreaterThan(orig, 0)
        XCTAssertEqual(ItbRuntime.setGOMAXPROCS(-3), orig, "a negative query must not change the value")
        XCTAssertEqual(ItbRuntime.setGOMAXPROCS(orig + 1), orig)
        XCTAssertEqual(ItbRuntime.setGOMAXPROCS(0), orig + 1)
        XCTAssertEqual(ItbRuntime.setGOMAXPROCS(orig), orig + 1)
    }

    /// A real path yields a non-empty file; an empty path with the
    /// environment fallback unset is rejected.
    func testWriteHeapProfile() throws {
        let path = FileManager.default.temporaryDirectory
            .appendingPathComponent("itb-swift-heap-\(getpid()).pprof").path
        try ItbRuntime.writeHeapProfile(path)
        let size = (try FileManager.default.attributesOfItem(atPath: path)[.size] as? Int) ?? 0
        XCTAssertGreaterThan(size, 0, "heap profile file is empty")
        try? FileManager.default.removeItem(atPath: path)

        unsetenv("ITB_MEMPROFILE")
        XCTAssertThrowsError(try ItbRuntime.writeHeapProfile("")) { error in
            XCTAssertEqual((error as? ItbError)?.status, .badInput)
        }
    }

    /// The length query sizes the buffer, and the filled buffer
    /// carries the tier count in slot 0 under the 1 + 5*T + 8 layout.
    func testPoolStats() throws {
        let len = ItbRuntime.poolStatsLen
        XCTAssertGreaterThanOrEqual(len, 9)
        let slots = try ItbRuntime.poolStats()
        XCTAssertEqual(slots.count, len)
        let tiers = Int(slots[0])
        XCTAssertGreaterThan(tiers, 0)
        XCTAssertEqual(1 + 5 * tiers + 8, len, "tier count does not match the slot count")
    }

    /// The shipped registry in canonical order, led by the Non-PRF
    /// inner primitive and carrying the harness default.
    func testHashNames() throws {
        let names = try hashNames()
        XCTAssertEqual(names.first, "aesitb128")
        XCTAssertTrue(names.contains("areion512"), "registry: \(names)")
    }

    /// The diagnostic is the only text an error carries, so no fixed
    /// buffer may bound it. The relay layer snapshots the library's
    /// text into an inline 2 KiB buffer and widens on the
    /// short-buffer return; a path the caller makes far longer than
    /// that is echoed back inside the os diagnostic, so an intact
    /// tail proves the widening branch ran rather than the text
    /// having been truncated at the boundary.
    func testWideDiagnosticSurvivesTheBoundary() {
        let tail = "zzzz-tail-marker"
        let longName = String(repeating: "a", count: 4000) + tail
        let path = "/tmp/" + longName
        XCTAssertThrowsError(try ItbRuntime.writeHeapProfile(path)) { error in
            guard let err = error as? ItbError else {
                return XCTFail("not an ItbError: \(error)")
            }
            XCTAssertGreaterThan(err.message.utf8.count, 2048,
                                 "diagnostic was clipped to the inline buffer: \(err.message.utf8.count) bytes")
            XCTAssertTrue(err.message.contains(tail),
                          "the end of the diagnostic was lost")
            XCTAssertTrue(err.message.hasSuffix("file name too long"),
                          "the reason at the end of the sentence was lost: \(err.message.suffix(40))")
        }
    }
}

File: bindings/fsharp/tests/Everanium.LibItb3.FSharp.Tests/RuntimeTests.fs
// Runtime diagnostics surface: GOMAXPROCS query / set / restore, the
// heap-profile writer, the pool-counter snapshot and its slot layout,
// and the hash-registry enumeration.

module Everanium.Itb3.FSharp.Tests.RuntimeTests

open System
open System.IO
open Xunit
open Everanium.Itb3.FSharp
open Everanium.Itb3.FSharp.Tests.TestSupport

[<Fact>]
let ``gomaxprocs query set restore`` () =
    let orig = Runtime.setGOMAXPROCS 0
    Assert.True(orig > 0)
    Assert.Equal(orig, Runtime.setGOMAXPROCS -3)
    Assert.Equal(orig, Runtime.setGOMAXPROCS(orig + 1))
    Assert.Equal(orig + 1, Runtime.setGOMAXPROCS 0)
    Assert.Equal(orig + 1, Runtime.setGOMAXPROCS orig)

[<Fact>]
let ``heap profile written and empty path rejected`` () =
    let dir =
        Path.Combine(Path.GetTempPath(), $"itb-loop-test-heap-fsharp-%d{Environment.ProcessId}")

    Directory.CreateDirectory dir |> ignore
    let path = Path.Combine(dir, "heap.prof")
    unwrap (Runtime.writeHeapProfile path)
    Assert.True(FileInfo(path).Length > 0L)
    Directory.Delete(dir, true)

    // The empty path falls back to ITB_MEMPROFILE inside libitb3; with
    // the variable clear there is nothing to fall back to.
    Environment.SetEnvironmentVariable("ITB_MEMPROFILE", null)

    match Runtime.writeHeapProfile "" with
    | Ok() -> failwith "expected an empty path to be rejected"
    | Error err -> Assert.Equal(Status.BadInput, err.Status)

[<Fact>]
let ``pool stats layout`` () =
    let len = Runtime.poolStatsLen ()
    Assert.True(len >= 9)
    let v = unwrap (Runtime.poolStats ())
    Assert.Equal(len, v.Length)
    let tiers = v[0]
    Assert.True(tiers > 0L)
    Assert.Equal(int64 len, 1L + 5L * tiers + 8L)

[<Fact>]
let ``hash names canonical`` () =
    let names = unwrap (Pipeline.hashNames ())
    Assert.Equal("aesitb128", List.head names)
    Assert.Contains("areion512", names)
